### PR TITLE
Use `download-rustc=false` global default, `if-unchanged` for tools and library profiles, and make `rust.debug-assertions=true` inhibit downloading CI rustc

### LINKED ISSUE #133132: `rust.debug-assertions = true` should be incompatible with `rust.download-rustc = true | 'if-unchanged'` atm
Before [Enable debug assertions on alt builds #131077](https://github.com/rust-lang/rust/pull/131077) lands, the alt rustc builds do *not* have rustc debug assertions enabled. However, currently `rust.download-rustc = true | "if-unchanged"` incorrectly *assumes* that the alt rustc has rustc debug assertions enabled (it does have *llvm* debug assertions enabled but not *rustc* debug assertions). This was the reason why #132789 was failing.

See [the investigation](https://rust-lang.zulipchat.com/#narrow/channel/326414-t-infra.2Fbootstrap/topic/When.20is.20rustc.20built.20with.20debug.20assertions.3F) and #133131 for details.

### PR DESCRIPTION
**Pull request by jieyouxu**
- Use `download-rustc = false` as global default.
    - Use `download-rustc = 'if-unchanged'` for tools and library profiles.
- Make `rust.debug-assertions = true` inhibit downloading CI rustc because alt rustc builds do not yet have rustc debug assertions enabled.

Fixes #133132.

cc discussions: https://rust-lang.zulipchat.com/#narrow/channel/326414-t-infra.2Fbootstrap/topic/Bootstrap.20breakage
compiler contributors poll: https://rust-lang.zulipchat.com/#narrow/channel/131828-t-compiler/topic/.60download-rustc.20.3D.20'if-unchanged'.60.20for.20.60compiler.60.20profile.3F/near/481877253
library contributors poll: https://rust-lang.zulipchat.com/#narrow/channel/219381-t-libs/topic/.60download-rustc.20.3D.20.22if-unchanged.22.60.20default.20for.20libs.20profile.3F/near/482607011
cc https://rust-lang.zulipchat.com/#narrow/channel/326414-t-infra.2Fbootstrap/topic/When.20is.20rustc.20built.20with.20debug.20assertions.3F

cc @MarcoIeni since you're working on improving CI job times, sorry, this will definitely regress some CI job times because we're probably lying to ourselves that CI rustc had debug assertions for some time 😅

cc @onur-ozkan for FYI, but since you're on vacation (sorry for the ping),

r? @Kobzol (I *think* you have a bit more context than other bootstrap reviewers?)

### CODE REVIEW

**Comment by jieyouxu** [comment]
I think the motivation for `download-rustc = 'if-unchanged'` for library profile is to prepare for the staging redesign, not 100% certain.

**Comment by jieyouxu** [comment]
Waiting for libs contributors poll to determine if we want to set the library profile to use `if-unchanged`.
@rustbot blocked

**Comment by jieyouxu** [comment]
This seems mostly uncontroversial to use `if-unchanged` for library profile after improved detection logic in #131831.

@rustbot ready

**Comment by jieyouxu** [comment]
Maybe r? @clubby789 (Onur is on vacation and Jakub/Mark/Albert is busy AFAIK)

File: src/bootstrap/src/core/config/config.rs
Comment (jieyouxu) [inline src/bootstrap/src/core/config/config.rs:1781]: I don't like this but this is the simplest thing I could think of without fixing our config layering

File: src/bootstrap/src/core/config/config.rs
Comment (clubby789) [inline src/bootstrap/src/core/config/config.rs:2810]: nit: `seamlessly`

**Review by clubby789 — APPROVED** [review]
Seems reasonable, r=me with typo fix

EDIT: There's also a leftover 'for' a couple of lines above

**Comment by jieyouxu** [comment]
Only change is typo fix ([diff](https://github.com/rust-lang/rust/compare/9f1063572c5de5a35a5d8c43da1a23a3484b99c5..1c4940e7bec357f0bd0bdad77d5513b65b5ee6ed)).

File: src/bootstrap/src/core/config/config.rs
Comment (jieyouxu) [reply src/bootstrap/src/core/config/config.rs]: Fixed.

**Comment by jieyouxu** [comment]
Yeah, like I suspected, this is going to be problematic.

@rustbot blocked (https://rust-lang.zulipchat.com/#narrow/channel/242791-t-infra/topic/CI.20jobs.20with.20.60--enable-debug-assertions.60.20.2B.20CI.20rustc)

**Comment by jieyouxu** [comment]
This will also need someone from infra to double-check... so r? @Mark-Simulacrum (or @Kobzol or infra)

**Comment by jieyouxu** [comment]
For `mingw-check`, AFAICT this is intended for use as a fast-rejection job, so for this specifically I set `NO_DEBUG_ASSERTIONS=1`, while for `x86_64-gnu-tools` I removed `FORCE_CI_RUSTC`.

**Comment by jieyouxu** [comment]
Hm this approach is problematic. I need to change the logic such that `rust.debug-assertions=true` inhibits `download-rustc` and force a fresh build.

**Comment by MarcoIeni** [comment]
> cc @MarcoIeni since you're working on improving CI job times, sorry, this will definitely regress some CI job times 

Thanks for letting me know! I will have a look at the job timings after this one is merged to make sure the auto build has reasonable timing 👍

File: src/bootstrap/src/core/config/config.rs
Comment (clubby789) [inline src/bootstrap/src/core/config/config.rs:1790]: nit: We're not very consistent about this within bootstrap, but IMO messages for user consumption should be `eprintln`'d so they don't get put into `grep` etc. This is mostly a preference thing and it's currently pretty 50/50 so up to you if you want to change this

File: src/bootstrap/src/core/config/config.rs
Comment (clubby789) [inline src/bootstrap/src/core/config/config.rs:2882]: nit: same as above `eprintln` note

**Comment by clubby789** [comment]
LGTM, r=me with or without comments addressed

**Comment by jieyouxu** [comment]
I'll adjust this PR tmrw, thanks for the review.

File: src/bootstrap/src/core/config/config.rs
Comment (jieyouxu) [reply src/bootstrap/src/core/config/config.rs]: Fixed.

File: src/bootstrap/src/core/config/config.rs
Comment (jieyouxu) [reply src/bootstrap/src/core/config/config.rs]: Modified to use this better wording and opened #133381 to track this.

File: src/bootstrap/src/core/config/config.rs
Comment (jieyouxu) [reply src/bootstrap/src/core/config/config.rs]: Fixed.

**Comment by jieyouxu** [comment]
### Changes since last review

- Swapped to `eprintln!` (both instances).
- Reworded comment and linked FIXME to new issue #133381.
- Pulled out change tracker entry into its own commit, and fixed the wording to more accurately reflect current logic.

**Comment by jieyouxu** [comment]
Review comments addressed, this has no other significant changes since last review, so
@bors r=@clubby789 rollup=never (significant bootstrap/CI changes)

**Comment by jieyouxu** [comment]
Actually I should wait for PR CI.
@bors r-

**Pull request merged**
